File: cpp/tests/python_scripts/estimators/regressors/rbf_test.py
# /*    Copyright (c) 2010-2018, Delft University of Technology
#  *    All rights reserved
#  *
#  *    This file is part of the Tudat. Redistribution and use in source and
#  *    binary forms, with or without modification, are permitted exclusively
#  *    under the terms of the Modified BSD license. You should have received
#  *    a copy of the license with this file. If not, please or visit:
#  *    http://tudat.tudelft.nl/LICENSE.
#  */

import random

from sympy import *
import numpy as np


if __name__ == '__main__':
  x1 = Symbol('x1')
  x2 = Symbol('x2')
  x3 = Symbol('x3')
  c1 = Symbol('c1')
  c2 = Symbol('c2')
  c3 = Symbol('c3')
  sigma = Symbol('sigma')

  random.seed(0)
  x = np.array([random.uniform(0,1), random.uniform(0,1), random.uniform(0,1)])
  print("x is ", x)

  c = np.array([random.uniform(0,1), random.uniform(0,1), random.uniform(0,1)])
  print("c is ", c)

  sigma_val = random.uniform(0,1)
  print("sigma is ", sigma_val)

  # Cubic
  cubic = Matrix([((x1 - c1)**2 + (x2 - c2)**2 + (x3 - c3)**2)**(3/2)])
  cubic_l = lambdify((x1, x2, x3, c1, c2, c3), cubic, 'numpy')
  print("Cubic evaluated at x and c is: ", cubic_l(x[0], x[1], x[2], c[0], c[1], c[2]))

  cubic_gradient = cubic.gradient([x1, x2, x3])
  cubic_gradient_l = lambdify((x1, x2, x3, c1, c2, c3), cubic_gradient, 'numpy')
  print("Cubic gradient evaluated at x and c is: ", cubic_gradient_l(x[0], x[1], x[2], c[0], c[1], c[2]))

  cubic_hessian = hessian(cubic, (x1, x2, x3))
  cubic_hessian_l = lambdify((x1, x2, x3, c1, c2, c3), cubic_hessian, 'numpy')
  print("Cubic Hessian evaluated at x and c is:\n", cubic_hessian_l(x[0], x[1], x[2], c[0], c[1], c[2]))

  # Gaussian
  gaussian = Matrix([exp(- ((x1 - c1)**2 + (x2 - c2)**2 + (x3 - c3)**2) / sigma**2)])
  gaussian_l = lambdify((x1, x2, x3, c1, c2, c3, sigma), gaussian, 'numpy')
  print("Gaussian evaluated at x, c and sigma is: ", gaussian_l(x[0], x[1], x[2], c[0], c[1], c[2], sigma_val))

  gaussian_gradient = gaussian.gradient([x1, x2, x3])
  gaussian_gradient_l = lambdify((x1, x2, x3, c1, c2, c3, sigma), gaussian_gradient, 'numpy')
  print("Gaussian gradient evaluated at x, c and sigma is: ", gaussian_gradient_l(x[0], x[1], x[2], c[0], c[1], c[2], sigma_val))

  gaussian_hessian = hessian(gaussian, (x1, x2, x3))
  gaussian_hessian_l = lambdify((x1, x2, x3, c1, c2, c3, sigma), gaussian_hessian, 'numpy')
  print("Gaussian Hessian evaluated at x, c and sigma is:\n", gaussian_hessian_l(x[0], x[1], x[2], c[0], c[1], c[2], sigma_val))

  #Expressions
  print("\nCubic:\n", cubic)

  print("\nCubic gradient:\n", cubic_gradient)

  print("\nCubic Hessian:\n")
  for line in cubic_hessian:
    print(line)

  print("\nGaussian:\n", gaussian)

  print("\nGaussian gradient:\n", gaussian_gradient)

  print("\nGaussian Hessian:\n")
  for line in gaussian_hessian:
    print(line)
  

File: cpp/tests/estimators/regressors/rbf_test.cpp
/*    Copyright (c) 2010-2018, Delft University of Technology
 *    All rights reserved
 *
 *    This file is part of the Tudat. Redistribution and use in source and
 *    binary forms, with or without modification, are permitted exclusively
 *    under the terms of the Modified BSD license. You should have received
 *    a copy of the license with this file. If not, please or visit:
 *    http://tudat.tudelft.nl/LICENSE.
 */

#include <vector>
#include <memory>
#include <cmath>

#include <iostream>

#include <Eigen/Core>

#include "tudat-learn/estimators/regressors/rbf.hpp"
 
int main() {
  std::cout.precision(9);

  // values obtained using cpp/tests/estimators/regressors/rbf_test.cpp
  std::vector< double > x({0.84442185, 0.7579544,  0.42057158});
  std::vector< double > c({0.25891675, 0.51127472, 0.40493414});

  double sigma = 0.7837985890347726;

  double cubic = 0.25670216;
  std::vector< double > cubic_jacobian({1.11633646, 0.47032472, 0.02981468});
  std::vector< double > cubic_hessian({
    3.52484826, 0.68177668, 0.04321898,
    0.68177668, 2.19386119, 0.01820863,
    0.04321898, 0.01820863, 1.90777552
  });

  double gaussian = 0.51815949;
  std::vector< double > gaussian_jacobian({-0.98767754, -0.41611931, -0.02637851});
  std::vector< double > gaussian_hessian({
    0.19575717,  0.79317606,  0.05028078,
    0.79317606, -1.35270747,  0.02118384,
    0.05028078,  0.02118384, -1.68553831
  });

  // Printing variables
  std::cout << "Variables:" << std::endl;
  std::cout << "x:" << std::endl;
  for(const auto &it : x)
    std::cout << it << ", ";
  std::cout << std::endl;
  std::cout << "c:" << std::endl;
  for(const auto &it : c)
    std::cout << it << ", ";
  std::cout << std::endl;
  std::cout << "sigma:\n" << sigma << std::endl;

  // Testing the CubicRBF
  std::cout << "\nTesting CubicRBF:" << std::endl;
  
  tudat_learn::CubicRBF<double> cubic_rbf;

  std::cout << "CubicRBF evaluated at x, c:" << std::endl;
  std::cout << cubic_rbf.eval(x, c) << std::endl;

  if(std::abs(cubic_rbf.eval(x, c) - cubic) > 1e-7 )
    return 1;


  auto cubic_jacobian_ptr = cubic_rbf.eval_jacobian(x, c);

  std::cout << "CubicRBF Jacobian evaluated at x, c:" << std::endl;
  for(auto i = 0; i < cubic_jacobian_ptr.get()->size(); ++i)
    std::cout << cubic_jacobian_ptr.get()->at(i) << ", ";
  std::cout << std::endl;

  if(cubic_jacobian_ptr.get()->size() != cubic_jacobian.size())
    return 1;

  for(auto i = 0; i < cubic_jacobian_ptr.get()->size(); ++i)
    if(std::abs(cubic_jacobian_ptr.get()->at(i) - cubic_jacobian.at(i)) > 1e-7 )
      return 1;  


  auto cubic_hessian_ptr = cubic_rbf.eval_hessian(x, c);

  std::cout << "CubicRBF Hessian evaluated at x, c:" << std::endl;
  for(auto i = 0; i < cubic_hessian_ptr.get()->size(); ++i)
    std::cout << cubic_hessian_ptr.get()->at(i) << ", ";
  std::cout << std::endl;

  if(cubic_hessian_ptr.get()->size() != cubic_hessian.size())
    return 1;

  for(auto i = 0; i < cubic_hessian_ptr.get()->size(); ++i)
    if(std::abs(cubic_hessian_ptr.get()->at(i) - cubic_hessian.at(i)) > 1e-7 )
      return 1;

  

  std::cout << "\nTesting GaussianRBF(sigma):" << std::endl;

  tudat_learn::GaussianRBF<double> gaussian_rbf(sigma);

  std::cout << "GaussianRBF(sigma) evaluated at x, c:" << std::endl;
  std::cout << gaussian_rbf.eval(x, c) << std::endl;

  if(std::abs(gaussian_rbf.eval(x, c) - gaussian) > 1e-7 )
    return 1;

  auto gaussian_jacobian_ptr = gaussian_rbf.eval_jacobian(x, c);

  std::cout << "GaussianRBF(sigma) Jacobian evaluated at x, c:" << std::endl;
  for(auto i = 0; i < gaussian_jacobian_ptr.get()->size(); ++i)
    std::cout << gaussian_jacobian_ptr.get()->at(i) << ", ";
  std::cout << std::endl;

  if(gaussian_jacobian_ptr.get()->size() != gaussian_jacobian.size())
    return 1;
    

  for(auto i = 0; i < gaussian_jacobian_ptr.get()->size(); ++i)
    if(std::abs(gaussian_jacobian_ptr.get()->at(i) - gaussian_jacobian.at(i)) > 1e-7 )
      return 1;  

  auto gaussian_hessian_ptr = gaussian_rbf.eval_hessian(x, c);

  std::cout << "GaussianRBF(sigma) Hessian evaluated at x, c:" << std::endl;
  for(auto i = 0; i < gaussian_hessian_ptr.get()->size(); ++i)
    std::cout << gaussian_hessian_ptr.get()->at(i) << ", ";
  std::cout << std::endl;

  if(gaussian_hessian_ptr.get()->size() != gaussian_hessian.size())
    return 1;

  for(auto i = 0; i < gaussian_hessian_ptr.get()->size(); ++i)
    if(std::abs(gaussian_hessian_ptr.get()->at(i) - gaussian_hessian.at(i)) > 1e-7 )
      return 1;


  return 0;
}


